api/pipeline_spec.proto

```diff
@@ -0,0 +1,420 @@
+      message Exec {
+        // The command of the exec program.
+        repeated string command = 2;
+        // The args of the exec program.
+        repeated string args = 3;
+      }
+      // This hook is invoked before caching check. It can change the properties
+      // of the execution and output artifacts before they are used to compute
+      // the cache key. The updated metadata will be passed into the main
+      // container entrypoint.
+      Exec pre_cache_check = 1;
+    }
+    // The lifecycle hooks of the container executor.
+    Lifecycle lifecycle = 4;
+  }
+
+  // The specification to import or reimport a new artifact to the pipeline.
+  message ImporterSpec {
+    // The URI of the artifact.
+    string artifact_uri = 1;
+
+    // The type of the artifact.
+    ArtifactTypeSchema type_schema = 2;
+
+    // The properties of the artifact.
+    map<string, ValueOrRuntimeParameter> properties = 3;
+
+    // The custom properties of the artifact.
+    map<string, ValueOrRuntimeParameter> custom_properties = 4;
+
+    // Whether or not import an artifact regardless it has been imported before.
+    bool reimport = 5;
+  }
+
+  // ResolverSpec is subject to change. Currently we only use enum to
+  // represent two of the currently available policies. We plan to introduce a
+  // flexible config to enable more sophisticated policies in the future.
+  // TODO(b/152230663): Support more flexibility for resolution logic.
+  message ResolverSpec {
+    enum ResolverPolicy {
+      RESOLVER_POLICY_UNSPECIFIED = 0;
+      // Within the models associated with the same pipeline context, always
+      // select the model with largest `last_update_time_since_epoch`.
+      LATEST_MODEL = 1;
+      // Within the models associated with the same pipeline context,
+      // select the latest Model that is blessed (represented by having a
+      // corresponding ModelBlessing artifact. The result will be consumed by
+      // components that need the latest model that is blessed such as
+      // Evaluator.
+      LATEST_BLESSED_MODEL = 2;
+    }
+    // TODO(b/152230663): Make this one of special resolver policies or generic
+    // ResolverConfig.
+    ResolverPolicy resolver_policy = 1;
+  }
+
+  // The specification of the executor.
+  message ExecutorSpec {
+    oneof spec {
+      // Starts a container.
+      PipelineContainerSpec container = 1;
+      // Import an artifact.
+      ImporterSpec importer = 2;
+      // Resolves an existing artifact.
+      ResolverSpec resolver = 3;
+    }
+  }
+  // Map from executor label to executor spec.
+  map<string, ExecutorSpec> executors = 1;
+}
+
+// Value is the value of the field.
+message Value {
+  oneof value {
+    // An integer value
+    int64 int_value = 1;
+    // A double value
+    double double_value = 2;
+    // A string value
+    string string_value = 3;
+  }
+}
+
+// The definition of a runtime artifact.
+message RuntimeArtifact {
+  // The name of an artifact.

```

Comment:
What is the "name of the artifact"?
What can it be?
How is it used?
How is it exposed?